### PR TITLE
[SPARK-19520][streaming] Do not encrypt data written to the WAL.

### PR DESCRIPTION
**Pull request by vanzin**
Spark's I/O encryption uses an ephemeral key for each driver instance.
So driver B cannot decrypt data written by driver A since it doesn't
have the correct key.

The write ahead log is used for recovery, thus needs to be readable by
a different driver. So it cannot be encrypted by Spark's I/O encryption
code.

The BlockManager APIs used by the WAL code to write the data automatically
encrypt data, so changes are needed so that callers can to opt out of
encryption.

Aside from that, the "putBytes" API in the BlockManager does not do
encryption, so a separate situation arised where the WAL would write
unencrypted data to the BM and, when those blocks were read, decryption
would fail. So the WAL code needs to ask the BM to encrypt that data
when encryption is enabled; this code is not optimal since it results
in a (temporary) second copy of the data block in memory, but should be
OK for now until a more performant solution is added. The non-encryption
case should not be affected.

Tested with new unit tests, and by running streaming apps that do
recovery using the WAL data with I/O encryption turned on.

### CODE REVIEW

**Comment by vanzin** [comment]
@tdas @zsxwing

**Comment by JoshRosen** [comment]
@liancheng FYI

**Comment by vanzin** [comment]
Kafka test error (hopefully unrelated?). Retest this please

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 1]
```diff
@@ -758,9 +761,33 @@ private[spark] class BlockManager(
       blockId: BlockId,
       bytes: ChunkedByteBuffer,
       level: StorageLevel,
-      tellMaster: Boolean = true): Boolean = {
+      tellMaster: Boolean = true,
+      encrypt: Boolean = false): Boolean = {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: I think its worth documenting this param.  At first I was going to suggest that it should be called `allowEncryption` like the other one, but I realize its more complicated than that.  Maybe something like

If true, the given bytes should be encrypted before they are stored.  Note that in most cases, the given bytes will *already* be encrypted if encryption is on.  An important exception to this is with the streaming WAL.  Since the WAL does not support encryption, those bytes are generated un-encrypted.  But we still encrypt those bytes before storing in the block manager.

Maybe too wordy but I think its worth documenting.

**Review by squito — COMMENTED** [review]
one minor suggestion for a doc, otherwise lgtm

File: streaming/src/main/scala/org/apache/spark/streaming/rdd/WriteAheadLogBackedBlockRDD.scala [hunk 2]
```diff
@@ -158,13 +158,16 @@ class WriteAheadLogBackedBlockRDD[T: ClassTag](
       logInfo(s"Read partition data of $this from write ahead log, record handle " +
         partition.walRecordHandle)
       if (storeInBlockManager) {
-        blockManager.putBytes(blockId, new ChunkedByteBuffer(dataRead.duplicate()), storageLevel)
+        blockManager.putBytes(blockId, new ChunkedByteBuffer(dataRead.duplicate()), storageLevel,
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/rdd/WriteAheadLogBackedBlockRDD.scala]: Why `encrypt` should be true here? In the following codes, it just reads the block using `maybeEncrypted = false`.

Comment (vanzin) [inline streaming/src/main/scala/org/apache/spark/streaming/rdd/WriteAheadLogBackedBlockRDD.scala]: This is explained in the summary.

The code that uses `maybeEncrypted = false` is deserializing data read from the WAL. This code is adding the block to the block manager, which later is read with `getBlockFromBlockManager` which calls `blockManager.get` which does decryption automatically.

Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/rdd/WriteAheadLogBackedBlockRDD.scala]: Got it

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 1 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: +1

**Comment by zsxwing** [comment]
Looks good.

**Comment by vanzin** [comment]
We can remove the `putBytes` workaround when I (or someone else) get around to fixing SPARK-19556.

For now, I'm merging this to master and 2.1.
